### PR TITLE
CM-619: IEnumerable, ICollection, IList and List - Which One to Use?

### PR DESCRIPTION
**Pull request by OldPen16**

### CODE REVIEW

File: collections-csharp/IEnumeralbleVsICollectionVsIListVsList/Tests/Test.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+using IEnumeralbleVsICollectionVsIListVsList;
+using NUnit.Framework;
+using System;
+
+namespace Tests
+{
+    public class Tests
```
Comment (karimkod) [inline collections-csharp/IEnumeralbleVsICollectionVsIListVsList/Tests/Test.cs:7]: Both The **file** and the class that contains the type should end with *UnitTest so I suggest to change this file to CollectionInterfacesUnitTest os something similar.

**Review by karimkod — CHANGES_REQUESTED** [review]

File: collections-csharp/IEnumeralbleVsICollectionVsIListVsList/IEnumeralbleVsICollectionVsIListVsList/Program.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+            var enu = new ImplementationOfIEnumerable();
+            Console.WriteLine($"IEnumerable:{enu.CountSpecialCharacters(specialCharacter)}\nIEnumerable With Yield:{ string.Join(",", enu.FilterNumbers(numbers))}");
+                
+            var icl = new ImplementationOfICollection();
+            Console.WriteLine($"ICollection:{icl.CountSpecialCharacters(specialCharacter)}");
+
+            //replacing CountSpecialCharactersArray with CountSpecialCharacters will end up in runtime exception
+            var ils = new ImplementationOfIList();
+            Console.WriteLine($"IList with List input:{ils.CountSpecialCharacters(specialCharacter)}\nIList with Array input:{ils.CountSpecialCharactersArray(arrayOfSpecialCharacter)}");
```
[the remark this answers is not in the document]
Comment (OldPen16) [reply collections-csharp/IEnumeralbleVsICollectionVsIListVsList/IEnumeralbleVsICollectionVsIListVsList/Program.cs:19]: Please can you add the article link here and point to the specific example?

File: collections-csharp/IEnumeralbleVsICollectionVsIListVsList/IEnumeralbleVsICollectionVsIListVsList/ImplementationOfIEnumerable.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            int total = 0;
+            foreach(int num in numbers)
+            {
+                if(num > 2)
+                {
+                    total += num;
+                    yield return total;
+                }   
+            }
+        }
```
[the remark this answers is not in the document]
Comment (OldPen16) [reply collections-csharp/IEnumeralbleVsICollectionVsIListVsList/IEnumeralbleVsICollectionVsIListVsList/ImplementationOfIEnumerable.cs:29]: I didn't find a yield keyword example in your shared links. Please can you share the link here and point it for my reference

File: collections-csharp/IEnumeralbleVsICollectionVsIListVsList/IEnumeralbleVsICollectionVsIListVsList/Program.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (karimkod) [reply collections-csharp/IEnumeralbleVsICollectionVsIListVsList/IEnumeralbleVsICollectionVsIListVsList/Program.cs:19]: Hello @OldPen16, sorry for I wasn't clear enough. 
I meant the article review and it's the point 6.b (just search in the docs 6.b and you will find my point about the array example).

File: collections-csharp/IEnumeralbleVsICollectionVsIListVsList/IEnumeralbleVsICollectionVsIListVsList/ImplementationOfIEnumerable.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (karimkod) [reply collections-csharp/IEnumeralbleVsICollectionVsIListVsList/IEnumeralbleVsICollectionVsIListVsList/ImplementationOfIEnumerable.cs:29]: The example of yield is in the article review docs and it's the point 2.a

**Comment by karimkod** [comment]
Hey @OldPen16,

1 - I don't see any tests in the test folder was this intentional or you missed to commit the test. 
It's important to have the test to verify the wanted behaviour of the code in the article. 

2 - There is a typo in all the code types : IEnumeralble should be IEnumerable.

**Pull request closed without merging**